### PR TITLE
Clarifications to SSO login/UIA

### PR DESCRIPTION
**Pull request by richvdh**
including a bunch of text about security.

### CODE REVIEW

File: specification/modules/sso_login.rst [hunk 1]
```diff
@@ -12,75 +12,164 @@
...
+implement a suitable endpoint. For example, for CAS authentication the
+homeserver should provide a means for the administrator to configure where the
+CAS server is and the REST endpoints which consume the ticket.
+
+Client login via SSO
+---------------------
+
+An overview of the process is as follows:
+
+0. The Matrix client calls |GET /login|_ to find the supported login
```
Comment (turt2live) [inline specification/modules/sso_login.rst:58]: Please can we start numbered lists at one?

File: specification/modules/sso_login.rst [hunk 2]
```diff
@@ -12,75 +12,164 @@
...
+   For example, consider a web-based client at ``https://client.example.com``,
+   which wants to initiate SSO login on the homeserver at ``server.example.org``.
+   It does this by storing the homeserver name in a query parameter for the
+   ``redirectUrl``: it redirects to
+   ``https://server.example.org/login/sso/redirect?redirectUrl=https://client.example.com?hs=server.example.org``.
+
+   An attacker could trick a victim into following a link to
+   ``https://server.example.org/login/sso/redirect?redirectUrl=https://client.example.com?hs=evil.org``,
+   which would result in the client sending a login token for the victim's
+   account to the attacker-controlled site ``evil.com``.
```
Comment (turt2live) [inline specification/modules/sso_login.rst:140]: example was .org
```suggestion
   account to the attacker-controlled site ``evil.org``.
```

File: specification/modules/sso_login.rst [hunk 3]
```diff
@@ -89,24 +178,169 @@ generate a URI for the SSO login page with any appropriate parameters.
...
+to distinguish which is taking place.
+
+The homeserver should validate the response from the SSO system: this may
+require additional calls to the authentication server, and/or may require
+checking a signature on the response.
+
+The homeserver then proceeds as follows:
+
+#. The homeserver MUST map the user details received from the authentication
+   server to a valid `Matrix user identifier <../index.html#user-identifiers>`_.
```
Comment (turt2live) [inline specification/modules/sso_login.rst:195]:
```suggestion
   server to a valid `Matrix user identifier <../appendices.html#user-identifiers>`_.
```

File: specification/modules/sso_login.rst [hunk 4]
```diff
@@ -89,24 +178,169 @@ generate a URI for the SSO login page with any appropriate parameters.
...
+   The guidance in `Mapping from other character sets
+   <../index.html#mapping-from-other-character-sets>`_ may be useful.
```
Comment (turt2live) [inline specification/modules/sso_login.rst:197]:
```suggestion
   <../appendices.html#mapping-from-other-character-sets>`_ may be useful.
```

File: specification/modules/sso_login.rst [hunk 5]
```diff
@@ -89,24 +178,169 @@ generate a URI for the SSO login page with any appropriate parameters.
...
+~~~~~~~~~~~~~~~~~~~~~~~
+
+1. Homeservers should ensure that login tokens are not sent to malicious
+   clients.
+
+   For example, consider a homeserver at ``server.example.org``. An attacker tricks
+   a victim into following a link to
+   ``https://server.example.org/login/sso/redirect?redirectUrl=https://evil.com``,
+   resulting in a login token being sent to the attacker-controlled site
+   ``evil.com``. This is a form of cross-site request forgery (CSRF).
```
Comment (turt2live) [inline specification/modules/sso_login.rst:226]: Ah, I see where the .com came from now - we should probably be consistent in the examples.

File: specification/modules/sso_login.rst [hunk 6]
```diff
@@ -89,24 +178,169 @@ generate a URI for the SSO login page with any appropriate parameters.
...
+1. Confirming the operation
+
+   The homeserver SHOULD confirm that the user is happy for the operation to go
+   ahead. The goal of the user-interactive authentication operation is to guard
+   against a compromised ``access_token`` being used to take over the user's
+   account. Simply redirecting the user to the SSO system is insufficient,
+   since they may not realise what is being asked of them, or the SSO system
+   may even confirm the authentication automatically.
+
+   For example, the homeserver might serve a page with words to the effact of:
```
Comment (turt2live) [inline specification/modules/sso_login.rst:319]:
```suggestion
   For example, the homeserver might serve a page with words to the effect of:
```

File: specification/modules/sso_login.rst [hunk 7]
```diff
@@ -89,24 +178,169 @@ generate a URI for the SSO login page with any appropriate parameters.
...
+     action, re-authenticate with single sign-on. If you did not expect this, your
+     account may be compromised!
 
-Once the homeserver has verified the user's identity with the SSO system, it
-MUST map the user ID to a valid `Matrix user identifier <../index.html#user-identifiers>`_.
-The guidance in `Mapping from other character sets
-<../index.html#mapping-from-other-character-sets>`_ may be useful.
+   This confirmation could take place before redirecting to the SSO
+   authentication page (when handling the
+   ``/_matrix/client/%CLIENT_MAJOR_VERSION%/auth/m.login.sso/fallback/web``
```
Comment (turt2live) [inline specification/modules/sso_login.rst:327]: fwiw we have a history of shortening URLs to after the version (ie: `/auth/m.login.sso/fallback/web`), largely for legibility.

File: specification/client_server_api.rst [hunk 8]
```diff
@@ -1,4 +1,4 @@
-.. Copyright 2016 OpenMarket Ltd
+.. Copyright 2016-2020 The Matrix.org Foundation C.I.C.
```
[the remark this answers is not in the document]
Comment (richvdh) [reply specification/client_server_api.rst:1]: yeah. CF https://github.com/matrix-org/synapse/issues/7565#issuecomment-639171606

File: specification/modules/sso_login.rst (hunk 1 above)
Comment (richvdh) [reply specification/modules/sso_login.rst:58]: we *can*, though in this case step zero felt very much like a preliminary step.

File: specification/modules/sso_login.rst [hunk 9]
```diff
@@ -12,75 +12,164 @@
...
+   cross-site request forgery (CSRF) attacks.
+
+   For example, consider a web-based client at ``https://client.example.com``,
+   which wants to initiate SSO login on the homeserver at ``server.example.org``.
+   It does this by storing the homeserver name in a query parameter for the
+   ``redirectUrl``: it redirects to
+   ``https://server.example.org/login/sso/redirect?redirectUrl=https://client.example.com?hs=server.example.org``.
+
+   An attacker could trick a victim into following a link to
+   ``https://server.example.org/login/sso/redirect?redirectUrl=https://client.example.com?hs=evil.org``,
```
Comment (richvdh) [inline specification/modules/sso_login.rst:138]:
```suggestion
   ``https://server.example.org/login/sso/redirect?redirectUrl=https://client.example.com?hs=evil.com``,
```

File: specification/modules/sso_login.rst (hunk 2 above)
Comment (richvdh) [reply specification/modules/sso_login.rst:140]: yeah. will change the example instead for consistency.

File: specification/modules/sso_login.rst (hunk 7 above)
Comment (richvdh) [reply specification/modules/sso_login.rst:327]: mmhmm. There was some inconsistency here and I went this way: do you think I should go the other?

File: specification/client_server_api.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (turt2live) [reply specification/client_server_api.rst:1]: reported as https://github.com/matrix-org/matrix-doc/issues/2613 to fix the rest

File: specification/modules/sso_login.rst (hunk 1 above)
Comment (turt2live) [reply specification/modules/sso_login.rst:58]: It's just a bit awkward to have things start at zero, as it implies that it's not a real step (at least imo).

File: specification/modules/sso_login.rst (hunk 7 above)
Comment (turt2live) [reply specification/modules/sso_login.rst:327]: I don't personally have a preference, just mentioning it in the event it felt annoying to have to do it every time.

File: specification/modules/sso_login.rst (hunk 1 above)
Comment (richvdh) [reply specification/modules/sso_login.rst:58]: that implication is correct, imo :)

File: specification/modules/sso_login.rst [hunk 10]
```diff
@@ -89,24 +178,169 @@ generate a URI for the SSO login page with any appropriate parameters.
...
+   The guidance in `Mapping from other character sets
+   <../index.html#mapping-from-other-character-sets>`_ may be useful.
+
+#. If the generated user identifier represents a new user, it should be
+   registered as a new user.
+
+#. The homeserver should generate a short-term login token. This is an opaque
+   token, suitable for use with the ``m.login.token`` type of the |/login|_
+   API. The lifetime of this token SHOULD be limited to around five
+   seconds.
```
Comment (KitsuneRal) [inline specification/modules/sso_login.rst:205]: This seems to assume relatively quick roundtrip between the client and the homeserver?

File: specification/modules/sso_login.rst [hunk 11]
```diff
@@ -89,24 +178,169 @@ generate a URI for the SSO login page with any appropriate parameters.
...
+
+#. The homeserver adds a query parameter of ``loginToken``, with the value of
+   the generated login token, to the ``redirectUrl`` given in the
+   ``/_matrix/client/%CLIENT_MAJOR_VERSION%/login/sso/redirect``
+   request. (Note: ``redirectURL`` may or may not include existing query
+   parameters. If it already includes one or more ``loginToken`` parameters,
+   they should be removed before adding the new one.)
+
+#. The homeserver redirects the user's browser to the URI thus built.
```
Comment (KitsuneRal) [inline specification/modules/sso_login.rst:214]:
```suggestion
#. The homeserver redirects the client to the URI thus built.
```

**Pull request merged**